src/cppsim/gate_named_npair.hpp

```diff
@@ -0,0 +1,139 @@
+    using UpdateFuncMpi = void (*)(UINT, UINT, UINT, CTYPE*, ITYPE, UINT);
+    UpdateFunc _update_func;
+    UpdateFunc _update_func_dm;
+    UpdateFuncGpu _update_func_gpu;
+    UpdateFuncMpi _update_func_mpi;
+    SparseComplexMatrix _matrix_element;
+    UINT _block_size;
+
+public:
+    explicit ClsNpairQubitGate(){};
+    /**
+     * \~japanese-en 量子状態を更新する
+     *
+     * @param state 更新する量子状態
+     */
+    virtual void update_quantum_state(QuantumStateBase* state) override {
+        if (state->is_state_vector()) {
+#ifdef _USE_GPU
+            if (state->get_device_name() == "gpu") {
+                throw NotImplementedException(
+                    this->_name + " gate is not Implemented");
+                // _update_func_gpu(this->_target_qubit_list[0].index(),
+                //     this->_target_qubit_list[1].index(), this->_block_size,
+                //     state->data(), state->dim, state->get_cuda_stream(),
+                //     state->device_number);
+            } else
+#endif
+#ifdef _USE_MPI
+                if (state->outer_qc > 0) {
+                _update_func_mpi(this->_target_qubit_list[0].index(),
+                    this->_target_qubit_list[1].index(), this->_block_size,
+                    state->data_c(), state->dim, state->inner_qc);
+            } else
+#endif
+            {
+                _update_func(this->_target_qubit_list[0].index(),
+                    this->_target_qubit_list[1].index(), this->_block_size,
+                    state->data_c(), state->dim);
+            }
+        } else {
+            throw NotImplementedException(
+                this->_name + " gate is not Implemented");
+            // _update_func_dm(this->_target_qubit_list[0].index(),
+            //     this->_target_qubit_list[1].index(), this->_block_size,
+            //     state->data_c(), state->dim);
+        }
+    };
+    /**
+     * \~japanese-en 自身のディープコピーを生成する
+     *
+     * @return 自身のディープコピー
+     */
+    virtual ClsNpairQubitGate* copy() const override {
+        return new ClsNpairQubitGate(*this);
+    };
+    /**
+     * \~japanese-en 自身のゲート行列をセットする
+     *
+     * @param matrix 行列をセットする変数の参照
+     */
+    virtual void set_matrix(ComplexMatrix& matrix) const override {
+        matrix = this->_matrix_element;
+    }
+
+    void FusedSWAPGateinit(
+        UINT target_qubit_index1, UINT target_qubit_index2, UINT block_size) {
+        this->_update_func = FusedSWAP_gate;
+        this->_update_func_dm = nullptr;  // not supported
+#ifdef _USE_GPU
+        this->_update_func_gpu = nullptr;  // not supported
+#endif
+#ifdef _USE_MPI
+        this->_update_func_mpi = FusedSWAP_gate_mpi;
+#endif
+        this->_name = "FusedSWAP";
+        this->_target_qubit_list.push_back(
+            TargetQubitInfo(target_qubit_index1, 0));
+        this->_target_qubit_list.push_back(
+            TargetQubitInfo(target_qubit_index2, 0));
```

Comment:
`target_qubit_list` should contain all the qubits to be updated.
```python
from qulacs import QuantumCircuit, QuantumState
from qulacs.circuit import QuantumCircuitOptimizer

circ = QuantumCircuit(4)
circ.add_H_gate(1)
circ.add_FusedSWAP_gate(0, 2, 2)
circ.add_H_gate(1)

optim = QuantumCircuitOptimizer()
optim.optimize_light(circ)

stat = QuantumState(4)
circ.update_quantum_state(stat)

print(stat.get_vector())
```
should print an unexcepted result.

Reply:
Thanks for pointing this out.
I will correct it.

Reply:
I fixed in commit 9e0559f